target_redshift/db_sync.py

```diff
@@ -653,109 +646,81 @@ def load_csv(self, s3_key, count, size_bytes, compression=False, manifest=False)
                             for pkey in names
                         ]
                     )
-                    sql = """
+                    sql = run(
+                        f"""
                     DELETE FROM {stage_table}
                     USING 
                       (SELECT {pkeys}, max(_sdc_sequence) AS "_sdc_sequence" FROM {stage_table} s GROUP BY {pkeys}) ss 
                     WHERE {join_condition} AND ss._sdc_sequence <> {stage_table}._sdc_sequence
-                    """.format(
-                        pkeys=pkeys,
-                        stage_table=stage_table,
-                        join_condition=join_condition,
+                    """
                     )
-                    self.logger.info("Running query: {}".format(sql))
-                    cur.execute(sql)
 
                     if not self.skip_updates:
-                        update_sql = """
+                        join_condition = " AND ".join(
+                            ["stage.{c} = target.{c}".format(c=pkey) for pkey in names]
+                        )
+
+                        run(
+                            f"""
                             UPDATE {target_table}
-                            SET _sys_end_time = stage._sys_updated_at
+                            SET _sys_end_time = COALESCE(stage._sys_updated_at, stage._sdc_deleted_at)
                             FROM {stage_table} stage
                             JOIN {target_table} target
                             ON {join_condition}
-                        """.format(
-                            target_table=target_table,
-                            stage_table=stage_table,
-                            join_condition=" AND ".join(
-                                [
-                                    "stage.{c} = target.{c}".format(c=pkey)
-                                    for pkey in names
-                                ]
-                            )
+                        """
                         )
 
-                        self.logger.info("Running query: {}".format(update_sql))
-                        cur.execute(update_sql)
+                        columns = ",".join([c["name"] for c in columns_with_trans])
+                        run(
+                            f"""
+                            INSERT INTO {history_table} ( {columns} )
+                            SELECT {columns}
+                            FROM {stage_table}
+                            WHERE _sdc_deleted_at IS NOT NULL
```

Comment:
This has to happen after the update.

I was hoping to not do another query but meh this should be fine. Basically insert into history when things are deleted to track changes.